README.md
# crelay
#### Controlling different relay cards for home automation with a Linux software

### About  
Ever bought a cute little USB relay card and wanted to use it on a Linux based device or computer? Chances are that your were out of luck because the Linux software support for the card provided by the manufacturer was non existent. Conrad, Sainsmart, Denkovi and friends are still completely ignoring the existence of this operating system. That's why I started this project.  

This software is intended to run on Linux systems to control USB relay cards from different manufacturers in a unified way. It provides several interfaces for controlling the relays locally or remotely via the network. The relays can be controlled by a human being via a device like smartphone or web browser, via a cloud service (e.g. IFTTT), or directly by an intelligent device as used in the Internet of Things.  
The software was designed with the following requirements in mind:  

 - simple, intuitive usage and interface
 - as little dependencies as possible (libraries, drivers, external programs)
 - runs on a variety of Linux distributions, different hardware platforms
 - lightweight, can run on simple devices
 - easily expandable (adding relay card types and user interfaces)

New relay cards support can be added by providing the cards driver code for detecting the card, reading and setting the relays.
Currently the following relay cards are supported:  

- [Conrad USB 4-channel relay card](http://www.conrad.de/ce/de/product/393905), 
  see [*Note 1*](https://github.com/ondrej1024/crelay#note-1-conrad-usb-4-channel-relay-card) below
- [Sainsmart USB 4/8-channel relay card](http://www.sainsmart.com/sainsmart-4-channel-5v-usb-relay-board-module-controller-for-automation-robotics.html), 
  see [*Note 2*](https://github.com/ondrej1024/crelay#note-2-sainsmart-usb-48-channel-relay-card) below
- [HID API compatible relay cards (1/2/4/8 channel)](http://www.ebay.com/itm/For-Smart-Home-5V-USB-Relay-2-Channel-Programmable-Computer-Control-/190950124351)
- [Sainsmart USB 16-channel relay control module](https://www.sainsmart.com/products/16-channel-usb-hid-programmable-control-relay-module)
- Generic GPIO controlled relays, 
  see [*Note 3*](https://github.com/ondrej1024/crelay#note-3-gpio-controlled-relays) below  

The used relay card is automatically detected. No complicated port or communication parameter settings required. Just plug in your card and play.  


The following picture shows a high level view on the modular software architecture.  

![Software architechture](screenshots/sw-architecture.png)
<br><br>

### Features
- Command line mode and daemon mode with Web GUI
- Automatic detection of relay card
- Reading of current relay states
- Setting of new relay states
- Single pulse generation on relay contact
- HTTP API for external clients (e.g. Smartphone/tablet apps)
- Multiple relay card type support
- Support for configuration file with custom parameters
- Multiple cards support (command line interface and HTTP API)
- IFTTT support (see [tutorial](tutorial_ifttt.md))
<br>

### Nice to have (wishlist)
- Integrated MQTT client (work in progress)
- Multiple cards support for Web GUI
- JSON format HTTP API
- Access control for Web GUI and HTTP API
- Programmable timers for relay actions  
<br>

### Screenshots

#### Web GUI
This is the Web GUI provided by the integrated web server. The web page dynamically adapts to the detected relay card as shown.

![Screenshot](screenshots/crelay-screenshot1.png "Web GUI with Conrad card connected")
<br>
----------
![Screenshot](screenshots/crelay-screenshot2.png "Web GUI with GPIO relays connected")
<br><br>

#### Command line interface
    $ crelay 
    crelay, version 0.14
    
    This utility provides a unified way of controlling different types of relay cards.
    Currently supported relay cards:
      - Conrad USB 4-channel relay card
      - Sainsmart USB 4/8-channel relay card
      - HID API compatible relay card
      - Sainsmart USB-HID 16-channel relay card
      - Generic GPIO relays
    The card which is detected first will be used, unless -s switch and a serial number is passed.
    
    The program can be run in interactive (command line) mode or in daemon mode with
    built-in web server.
    
    Interactive mode:
        crelay [-s <serial number>] -i | [<relay number>] [ON|OFF]
    
           The state of any relay can be read or it can be changed to a new state.
           If only the relay number is provided then the current state is returned,
           otherwise the relays state is set to the new value provided as second parameter.
           The USB communication port is auto detected. The first compatible device
           found will be used, unless -s switch and a serial number is passed.
    
    Daemon mode:
        crelay -d [<relay1_label> [<relay2_label> [<relay3_label> [<relay4_label>]]]] 
    
           In daemon mode the built-in web server will be started and the relays
           can be completely controlled via a Web browser GUI or HTTP API.
           Optionally a personal label for each relay can be supplied which will
           be displayed next to the relay name on the web page.
    
           To access the web interface point your Web browser to the following address:
           http://<my-ip-address>:8000
    
           To use the HTTP API send a POST or GET request from the client to this URL:
           http://<my-ip-address>:8000/gpio
<br>  

### HTTP API
An HTTP API is provided to access the server from external clients. This API is compatible with the [PiRelay Android app](https://play.google.com/store/apps/details?id=com.jasonfindlay.pirelaypro). Therefore the app can be used on your Android phone to control *crelay* remotely.  

*Additionally to the existing API a universal Json format based API should be imlpemented. This would make it easier for third party clients to send API requests and parse the response. Any volunteer who wants to contribute is very welcome.*

- API url:  
<pre><i>ip_address[:port]</i>/gpio</pre>  

- Method:  
<pre>POST or GET</pre>  

- Reading relay states  
Required Parameter: none  

- Setting relay state  
Required Parameter: <pre>pin=[1|2|3 ...], status=[0|1|2] where 0=off 1=on 2=pulse</pre>
Optional Parameter: <pre>serial=*serial_number*</pre>

- Response from server:  
<pre>
Relay 1:[0|1]
Relay 2:[0|1]
Relay 3:[0|1]
Relay 4:[0|1]
</pre>  
<br>

### Installation from source
The installation procedure is usually perfomed directly on the target system. Therefore a C compiler and friends should already be installed. Otherwise a cross compilation environment needs to be setup on a PC (this is not described here).  

* Install dependencies (package names may vary depending on your distribution):

If you have a Debian, Ubuntu or derivative distribution:
<pre>
    apt-get install libftdi1 libftdi-dev libhidapi-libusb0 libhidapi-dev libusb-1.0-0 libusb-1.0-0-dev
</pre>

If you have OpenSUSE or a derivative distribution:
<pre>
    zypper install libftdi1-devel libftdi0-devel libhidapi-libusb0 libhidapi-devel libusb-1_0-devel libusb-1_0-0 libftdi1 libftdi1-2
</pre>

* Build dependencies from source (optional):  
On some not so recent Linux distributions (like Debian Weezy) the HIDAPI library is not yet packaged, so it needs to be build from the source code. Follow these instructions to do that:
<pre>
    apt-get install libudev-dev libusb-1.0-0-dev
    git clone git://github.com/signal11/hidapi.git
    ./bootstrap
    ./configure
    make
    sudo make install
</pre>

* Clone crelay git repository :  
<pre>
    git clone https://github.com/ondrej1024/crelay
    cd crelay
</pre>

* Alternatively get latest source code version :  
<pre>
    wget https://github.com/ondrej1024/crelay/archive/master.zip
    unzip master.zip
    cd crelay-master
</pre>

* Build and install :  
<pre>
    cd src
    make [DRV_CONRAD=n] [DRV_SAINSMART=n] [DRV_HIDAPI=n]
    sudo make install
</pre>
<i>Note:</i> Optionally, you can exclude specific relay card drivers (and their dependencies) from the build, if you don't need them. To do this, specify the driver name as parameter of the "make" command as shown above.  

<br>

### Installation of prebuilt binaries
To save you the hassle of building crelay from source, prebuild binaries are provided for selected architectures. Just save the binary for your architecture on your host system and execute it.  

The binaries can be downloaded from the [Latest release page](https://github.com/ondrej1024/crelay/releases/latest).

<i>Note:</i> The binaries use shared librabries, so you might need to install the needed libs to your system, if not already done previously:  
<pre>
    apt-get install libftdi1 libhidapi-libusb0 libusb-1.0-0
</pre>  
<br>

### Distribution specific packages
Currently, there are official OpenWRT packages available for different architectures from the developement snapshot repositories. They can be downloaded here:  

    https://downloads.openwrt.org/snapshots/trunk/[arch]/generic/packages/packages/crelay<ver>_<arch>.ipk

If anyone wants to build packages for other Linux distributions, please feel free to do so and share it here.  

<br>

### Configuration
When running in daemon mode. the parameters for *crelay* can be customized via the configuration file `crelay.conf` which should be placed in the `/etc/` system folder. An example file is provided together with the programs source code in the `conf/` folder.

<pre>
################################################
#
# crelay config file
#
# This file is read by crelay in daemon mode
# from /etc/crelay.conf
#
################################################

# HTTP server parameters
################################################
[HTTP server]
server_iface = 0.0.0.0    # listen interface IP address
#server_iface = 127.0.0.1 # to listen on localhost only
server_port  = 8000       # listen port
relay1_label = Device 1   # label for relay 1
relay2_label = Device 2   # label for relay 2
relay3_label = Device 3   # label for relay 3
relay4_label = Device 4   # label for relay 4
relay5_label = Device 5   # label for relay 5
relay6_label = Device 6   # label for relay 6
relay7_label = Device 7   # label for relay 7
relay8_label = Device 8   # label for relay 8
    
# GPIO driver parameters
################################################
[GPIO drv]
#num_relays = 8    # Number of GPIOs connected to relays (1 to 8)
#active_value = 1       # 1: active high, 0 active low
#relay1_gpio_pin = 17   # GPIO pin for relay 1 (17 for RPi GPIO0)
#relay2_gpio_pin = 18   # GPIO pin for relay 2 (18 for RPi GPIO1)
#relay3_gpio_pin = 27   # GPIO pin for relay 3 (27 for RPi GPIO2)
#relay4_gpio_pin = 22   # GPIO pin for relay 4 (22 for RPi GPIO3)
#relay5_gpio_pin = 23   # GPIO pin for relay 5 (23 for RPi GPIO4)
#relay6_gpio_pin = 24   # GPIO pin for relay 6 (24 for RPi GPIO5)
#relay7_gpio_pin = 25   # GPIO pin for relay 7 (25 for RPi GPIO6)
#relay8_gpio_pin = 4    # GPIO pin for relay 8 ( 4 for RPi GPIO7)
    
# Sainsmart driver parameters
################################################
[Sainsmart drv]
num_relays = 4   # Number of relays on the Sainsmart card (4 or 8)
</pre>

<br>

### Adding new relay card drivers
The modular architecture of *crelay* makes it possible to easily add new relay card drivers.  
See example files `relay_drv_sample.c` and `relay_drv_sample.h` in the src directory for details on how to write your own low level driver functions.  
<br>

### Service autostart  

In order to start `crelay` as daemon automatically on system boot it can be configured as `systemd` service. The service file is provided here:

`systemd/crelay.service`

To enable the service, run the following commands:

```
cp systemd/crelay.service /etc/systemd/system/
systemctl daemon-reload
systemctl enable crelay.service
```

Start/stop the service manually with these commands:

```
systemctl start crelay.service
systemctl stop crelay.service
```



### Credits
The support for the different relay cards in *crelay* has only been possible thanks to the valuable contributions of the following people who reverse engineered the communication protocols of the cards and provided the protocol specifications.

* [Dominic Sacré](https://github.com/dsacre) who discovered the communication protocol of the Conrad USB card

* [Darryl Bond](https://github.com/darrylb123), who discovered the communication protocol of the HID API cards

* [Steve Crow](https://github.com/stav09), who discovered the communication protocol of the Sainsmart USB cards

* [Kevin Hilman](https://github.com/khilman), who implemented and tested the support for the Sainsmart 16-channel control module

* [Andrew Lunn](https://github.com/lunn), who contributed cleanup patches

* [Andrey Shevtsov](https://github.com/sqlwristband), who contributed the initial multiple cards implementation

* [Alberto Bursi](https://github.com/bobafetthotmail), who contributed documentation for OpenSUSE build support

* [Derek Atkins](https://github.com/derekatkins), who contributed multiple cards handling for the Sainsmart 16 Channel HID controller and HTTP API

* [Paweł Dembicki](https://github.com/CHKDSK88), who contributed to the GPIO driver

  <br>  

### Notes

##### <i>Note 1 (Conrad USB 4-channel relay card)</i>:
The relay card software provided by Conrad is Windows only and uses a binary runtime DLL which implements the communication protocol between the host computer and the card. Thanks to a raspberrypi.org forum member, the communication protocol was discovered and made public. This made it possible to develop an open source driver for the Conrad card which can run on any Linux distribution.

Earlier versions of this program needed the cp210x kernel driver for the Silabs CP2104 chip with GPIO support. The official in-kernel cp210x driver does not  support GPIO operations. Therefore the Silabs driver from their home page needed to be used:
http://www.silabs.com/products/mcu/pages/usbtouartbridgevcpdrivers.aspx

The current version of *crelay* uses libusb to talk directly to the CP2104 on the Conrad card, therefore the cp210x kernel driver is **not needed** anymore.  
<br>  

##### <i>Note 2 (Sainsmart USB 4/8-channel relay card)</i>:
The Sainsmart card uses the FTDI FT245RL chip. This chip is controlled directly through the open source libFTDI library. No Kernel driver is needed. However on most Linux distributions, the *ftdi_sio* serial driver is automatically loaded when the FT245RL chip is detected. In order to grant the *crelay* software access to the card, the default driver needs to be unloaded:  

    rmmod ftdi_sio

To prevent automatic loading of the driver, add the following line to /etc/modprobe.d/blacklist.conf:  

    blacklist ftdi_sio

Both 4 and 8 channel versions are supported. However, there seems to be no way to automatically detect which version of the card is used. Therefore the number of relay channles can be configured in the configuration file.  

<br>  

##### <i>Note 3 (GPIO controlled relays)</i>:
Since GPIO pin configuration is strictly device specific, the generic GPIO mode is disabled by default and can only be used in daemon mode. In order to enable it, the specific GPIO pins used as relay control lines have to be specified in the configuration file, `[GPIO drv]` section.  



HISTORY
Version history
===============

0.1: 23/12/2013
- First release, interactive (command line) mode only

0.2: 18/01/2014
- Added integrated web server which provides a Web GUI in daemon mode

0.3: 22/01/2014
- Added HTTP API which can be used by external clients (e.g. Android app)

0.4: 20/02/2014
- Changed code structure to allow easy inclusion of new relay card support

0.5: 26/02/2014
- Added support for relays connected via GPIO pins

0.6: 27/01/2015
- Added support for Sainsmart USB 4-channel relay card

0.7: 10/01/2015
- Added support for HID API compatible relay cards
- Added "make" parameters to exclude specific relay card drivers and their
  dependencies from the build
- Added copyright note and license text to source files
- Added sample driver source files

0.7.1: 22/04/2015
- Added init script for autostarting crelay in daemon mode
- Modified GPIO driver to adapt to new Kernel behaviour

0.8: 14/08/2015
- New implementation for the Conrad card driver, now uses libusb and
  therefore completely removes the need for the patched cp210x kernel driver
  
0.9: 20/08/2015
- Added parameter handling in daemon mode via configuration files
- Added new parameters for daemon listen interface and port
- Added new parameters for GPIO pin definitions
- Added new parameter for number of relays on Sainsmart card (4/8)
- Added detection for number of relays on HID API compatible relay card
- Added logging daemon messages to syslog

0.10: 11/04/2016
- Added support for SainSmart 16 Channel controller USB HID Programmable Control Relay Module
- Cleanup patches

0.10.1: 30/08/2016 (Bugfix release)
- Fixed compatibility issue with PiRelay app
- Added "-D" option for daemon mode ("-D" sends crelay to backgound, "-d" leaves crelay in foregraound)
- Modified file stream handling as temporary workaround for web server connection issue on some systems (OpenWRT)
- Added LDFLAGS parameter to linger command in Makefile (needed by OpenWRT)

0.11: 30/12/2016
- Added multiple cards support for command line ("-s" option to specify cards serial number)
- Created C library version of crelay for inclusion in third party applications

0.12: 03/11/2017
- Handle GPIO relays with inverse logic (active low)

0.13: 20/08/2018
- Added error handling for web socket operations
- Added monit script for crelay process supervision in daemon mode
- Added build dependencies for OpenSUSE to readme 
- Added support for multiple SainSmart 16-channel HID relay cards

0.14: 13/01/2019
- Added multiple cards support to HTTP API (new "serial" parameter)
- Reworked WebGUI (replaced submit buttons with Javascript toggle switches)

0.14.1: 13/12/2021
- Fixed segfault caused by Sainsmart driver
- Add systemd service file


src/relay_drv.h
/******************************************************************************
 * 
 * Relay card control utility: Relay driver generic file
 * 
 * Description:
 *   This software is used to controls different type of relays cards.
 *   This file contains the declaration of the generic functions.
 * 
 * Author:
 *   Ondrej Wisniewski (ondrej.wisniewski *at* gmail.com)
 *
 * Last modified:
 *   19/08/2015
 *
 * Copyright 2015, Ondrej Wisniewski 
 * 
 * This file is part of crelay.
 * 
 * crelay is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with crelay.  If not, see <http://www.gnu.org/licenses/>.
 * 
 *****************************************************************************/ 
 
#ifndef relay_drv_h
#define relay_drv_h

/* Conrad 4 channel USB relay card */
#define CONRAD_4CHANNEL_USB_NAME       "Conrad USB 4-channel relay card"
#define CONRAD_4CHANNEL_USB_NUM_RELAYS 4

/* Sainsmart 4/8 channel USB relay card */
#define SAINSMART_USB_NAME             "Sainsmart USB 4/8-channel relay card"
#define SAINSMART_USB_NUM_RELAYS       8

/* HID API compatibe x channel relay card */
#define HID_API_RELAY_NAME             "HID API compatible relay card"
#define HID_API_NUM_RELAYS             8

/* Sainsmart HID API compatibe 16-channel relay card */
#define SAINSMART16_USB_NAME          "Sainsmart USB-HID 16-channel relay card"
#define SAINSMART16_USB_NUM_RELAYS     16

/* Generic GPIO connected relay cards */
#define GENERIC_GPIO_NAME              "Generic GPIO relays"
#define GENERIC_GPIO_NUM_RELAYS        8


#define FIRST_RELAY    1
#define MAX_NUM_RELAYS 16
#define MAX_RELAY_CARD_NAME_LEN 40
#define MAX_COM_PORT_NAME_LEN 32
#define MAX_SERIAL_LEN 32


typedef enum
{
   NO_RELAY_TYPE=0,
   
#ifdef DRV_CONRAD
   CONRAD_4CHANNEL_USB_RELAY_TYPE, /* Conrad usb 4-channel relay card */
#endif
#ifdef DRV_SAINSMART
   SAINSMART_USB_RELAY_TYPE,       /* Sainsmart usb 4/8-channel relay card */
#endif
#ifdef DRV_HIDAPI
   HID_API_RELAY_TYPE,             /* HID API compatible relay card */
#endif
#ifdef DRV_SAINSMART16
   SAINSMART16_USB_RELAY_TYPE,     /* Sainsmart USB-HID relay card */
#endif
   
   /* Add other relay types here */
   
#ifndef BUILD_LIB
   GENERIC_GPIO_RELAY_TYPE,        /* Relays connected directly via GPIO pins */
#endif
   LAST_RELAY_TYPE
} relay_type_t;

typedef enum 
{
   OFF=0,
   ON,
   PULSE,
   INVALID
}
relay_state_t;

typedef struct relay_info
{
   relay_type_t relay_type;
   char         serial[MAX_SERIAL_LEN]; 
   struct relay_info *next;
} 
relay_info_t;

typedef struct
{
   int (*detect_relay_card_fun)(char*, uint8_t*, char*, relay_info_t **); /* function to detect the relay card */
   int (*get_relay_fun)(char*, uint8_t, relay_state_t*, char*); /* function to get the current relay state */
   int (*set_relay_fun)(char*, uint8_t, relay_state_t, char*);  /* function to set the new relay state */
   char *card_name;                                           /* card name string */
}
relay_data_t;



/**********************************************************
 * Function crelay_detect_all_relay_cards()
 * 
 * Description: Detect all relay cards
 * 
 * Parameters: relay_info(out)- pointer to list of 
 *                              relays info struct
 * 
 * Return:  0 - success
 *         -1 - fail, no relay card found
 *********************************************************/
int crelay_detect_all_relay_cards(relay_info_t** relay_info);

/**********************************************************
 * Function crelay_detect_relay_card()
 * 
 * Description: Detect the relay card
 * 
 * Parameters: portname (out) - pointer to a string where
 *                              the detected com port will
 *                              be stored
 *             num_relays(out)- pointer to number of relays
 * 
 * Return:  0 - success
 *         -1 - fail, no relay card found
 *********************************************************/
int crelay_detect_relay_card(char* portname, uint8_t* num_relays, char* serial, relay_info_t** relay_info);

/**********************************************************
 * Function crelay_get_relay()
 * 
 * Description: Get the current relay state
 * 
 * Parameters: portname (in)     - communication port
 *             relay (in)        - relay number
 *             relay_state (out) - current relay state
 * 
 * Return:   o - success
 *          -1 - fail
 *********************************************************/
int crelay_get_relay(char* portname, uint8_t relay, relay_state_t* relay_state, char* serial);

/**********************************************************
 * Function crelay_set_relay()
 * 
 * Description: Set new relay state
 * 
 * Parameters: portname (in)     - communication port
 *             relay (in)        - relay number
 *             relay_state (in)  - current relay state
 * 
 * Return:   o - success
 *          -1 - fail
 *********************************************************/
int crelay_set_relay(char* portname, uint8_t relay, relay_state_t relay_state, char* serial);

/**********************************************************
 * Function crelay_get_relay_card_type()
 * 
 * Description: Get the detected relay type
 * 
 * Parameters: none
 * 
 * Return: relay type
 *********************************************************/
relay_type_t crelay_get_relay_card_type();

/**********************************************************
 * Function crelay_get_relay_card_name()
 * 
 * Description: Get the detected relay card name
 * 
 * Parameters: none
 * 
 * Return: relay card name
 *********************************************************/
int crelay_get_relay_card_name(relay_type_t rtype, char* card_name);

#endif

